src/python/grpcio_testing/grpc_testing/_time.py

```diff
@@ -0,0 +1,223 @@
+                state.behaviors_to_times.pop(mature_behavior)
+            mature_behaviors.extend(earliest_mature_behaviors)
+        else:
+            earliest_immature_time = earliest_time
+            break
+    else:
+        earliest_immature_time = None
+    return _Delta(mature_behaviors, earliest_mature_time,
+                  earliest_immature_time)
+
+
+class _Future(grpc.Future):
+
+    def __init__(self, state, behavior):
+        self._state = state
+        self._behavior = behavior
+        self._cancelled = False
+
+    def cancel(self):
+        with self._state.condition:
+            time = self._state.behaviors_to_times.pop(self._behavior, None)
+            if time is None:
+                return False
+            else:
+                self._state.times_to_behaviors[time].remove(self._behavior)
+                if not self._state.times_to_behaviors[time]:
+                    self._state.times_to_behaviors.pop(time)
+                    self._state.condition.notify_all()
+                self._cancelled = True
+                return True
+
+    def cancelled(self):
+        with self._state.condition:
+            return self._cancelled
+
+    def running(self):
+        raise NotImplementedError()
+
+    def done(self):
+        raise NotImplementedError()
+
+    def result(self, timeout=None):
+        raise NotImplementedError()
+
+    def exception(self, timeout=None):
+        raise NotImplementedError()
+
+    def traceback(self, timeout=None):
+        raise NotImplementedError()
+
+    def add_done_callback(self, fn):
+        raise NotImplementedError()
+
+
+class RealTime(grpc_testing.Time):
+
+    def __init__(self):
+        self._state = _State()
+        self._active = False
+        self._calling = None
+
+    def _activity(self):
+        while True:
+            with self._state.condition:
+                while True:
+                    now = _time.time()
+                    delta = _process(self._state, now)
+                    self._state.condition.notify_all()
+                    if delta.mature_behaviors:
+                        self._calling = delta.earliest_mature_time
+                        break
+                    self._calling = None
+                    if delta.earliest_immature_time is None:
+                        self._active = False
+                        return
+                    else:
+                        timeout = max(0, delta.earliest_immature_time - now)
+                        self._state.condition.wait(timeout=timeout)
+            _call(delta.mature_behaviors)
+
+    def _ensure_called_through(self, time):
+        with self._state.condition:
+            while ((self._state.times_to_behaviors and
+                    min(self._state.times_to_behaviors) < time) or
+                   (self._calling is not None and self._calling < time)):
+                self._state.condition.wait()
+
+    def _call_at(self, behavior, time):
```

Comment:
A system should never in the implementation of one part of its public API make casual use of another part of its public API. See "Effective Java" (Second Edition) Item 17, [_The End Of Object Inheritance And The Beginning Of A New Modularity_](https://www.youtube.com/watch?v=3MNVP9-hglc), and... lots of other software engineering guidance over the years.

I write a lot of systems that look like
```
def _A()
  <implementation>

def A()
  return _A()

def B()
  <implementation that uses _A>
```
and few that look like
```
def A()
  <implementation>

def B()
  <implementation that uses A>
```
.